### ISSUE TITLE
PyISY crashes on bogus (non-UTF8) data from the ISY

### ISSUE THREAD

**Issue by derekatkins** [automicus/PyISY#124]
I had a bogus zwave node in the ISY that got added automagically (even though I told the ISY not to), and when it reported, the formatted text was in non-UTF8.  This causes PyISY to throw a parsing error.  It should, instead, have just ignored the error and continued on.

See https://github.com/home-assistant/core/issues/40134 for more information.

**Comment by bdraco** [comment]
Potential solution v3



Try changing
```
  File "/usr/local/lib/python3.8/site-packages/pyisy/connection.py", line 149, in request
````


```
    results = await res.text()
```

to

```
    results = await res.text(errors='ignore')
```

**Comment by bdraco** [comment]
Potential solution v2

```
  File "/usr/local/lib/python3.8/site-packages/pyisy/eventreader.py", line 53, in read_events
```

```
    events.append(body.decode("utf-8"))
```

to

```
    events.append(body.decode("utf-8", 'ignore'))
```

**Comment by derekatkins** [comment]
Unfortunately I cleared out the bogus node on my ISY so I cannot test these changes.  Also, as I'm running HassOS, I don't seem to have access to /usr/local/lib/python3.8/  (or the directory is different from the SSH Add-On container)

**Comment by derekatkins** [comment]
I am happy to supply you (privately) a status.xml that has the bogus data.

**Comment by shbatm** [comment]
@derekatkins please send it to my email on my profile as well as nodes.xml if you have it. I can run it against @bdraco's changes tomorrow when I get a chance using an ISY simulator. They look like they'll take care of the issue. 

From skimming the HA issue it's definitely on the get_status call and this is not the first time this bad UTF-8 data from the ISY has surfaced... Although before (in V1) it was just translated into the funky characters like you saw in your xml.

**Comment by derekatkins** [comment]
@shbatm should be in your inbox now.  Thanks!

**Comment by shbatm** [comment]
Fix has been pushed for V2 and will be included in next release. Fix for V3 is outstanding in #126; works for initial loading, but have not been able to test on live event stream yet.

**Comment by derekatkins** [comment]
It partially works.  At 10:35:02 I stopped receiving sensor data from one my of Z-Wave multi-sensors (ZW003_1).  Here's what I see in the log around that time:
```txt
...
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_1': control='CLIHUM', value='66', prec='0', uom='22', formatted='66%')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_1': control='CLITEMP', value='386', prec='1', uom='17', formatted='38.6°F')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_1': control='BATLVL', value='100', prec='0', uom='51', formatted='100%')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_104
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW002_308': control='ALARM', value='0', prec='0', uom='25', formatted='0')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW004_1
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_172
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_157
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_173
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: 52 E5 26 1
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: 52 E5 26 2
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_153
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_155
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW002_157
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW002_155
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW002_104
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_308': control='ALARM', value='0', prec='0', uom='25', formatted='0')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_144': control='PULSCNT', value='19530108', prec='0', uom='55', formatted='19530108')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW009_1
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: ZW003_143
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('51 1A FB 1': control='RR', value='28', prec='0', uom='57', formatted='28')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('51 1A FB 1': control='OL', value='255', prec='0', uom='100', formatted='100%')
2020-09-19 10:35:02 DEBUG (MainThread) [pyisy] ISY Updated Node: 51 1A FB 1
2020-09-19 10:35:03 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('50 BE 48 1': control='RR', value='28', prec='0', uom='57', formatted='28')
2020-09-19 10:35:03 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('50 BE 48 1': control='OL', value='255', prec='0', uom='100', formatted='100%')
2020-09-19 10:35:03 DEBUG (MainThread) [pyisy] ISY Updated Node: 50 BE 48 1
2020-09-19 10:35:03 ERROR (MainThread) [pyisy.events.websocket] Error during receive Invalid UTF-8 text message
2020-09-19 10:35:03 WARNING (MainThread) [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
2020-09-19 10:35:03 DEBUG (MainThread) [pyisy.events.websocket] Stopping websocket connection.
2020-09-19 10:35:03 INFO (MainThread) [pyisy.events.websocket] PyISY attempting stream reconnect in 0.01s.
2020-09-19 10:35:03 DEBUG (MainThread) [pyisy.events.websocket] Starting websocket connection.
2020-09-19 10:35:04 DEBUG (MainThread) [pyisy.events.websocket] Successfully connected to websocket.
2020-09-19 10:35:04 DEBUG (MainThread) [pyisy.events.websocket] ISY HEARTBEAT: 2020-09-19T10:35:04.505854
2020-09-19 10:35:04 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('50 F E1 1': control='RR', value='28', prec='0', uom='57', formatted='0.5 seconds')
2020-09-19 10:35:04 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('50 F E1 1': control='OL', value='255', prec='0', uom='100', formatted='100%')
...
```
But after this point I never get anything back from ZW003_1, even though I was seeing messages in the ISY Event Log.  Even worse, my ZW010_1 stopped talking to HA yesterday at 19:19:29, which also closely corresponds to a similar error:
```txt
2020-09-18 19:19:29 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW010_1': control='CLIHUM', value='59', prec='0', uom='22', formatted='59%')
2020-09-18 19:19:29 DEBUG (MainThread) [pyisy] ISY Updated Program: 003C
2020-09-18 19:19:30 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW002_1': control='LUMIN', value='0', prec='0', uom='36', formatted='0 lux')
2020-09-18 19:19:31 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_1': control='CLIHUM', value='73', prec='0', uom='22', formatted='73%')
2020-09-18 19:19:31 DEBUG (MainThread) [pyisy] ISY Updated Program: 003D
2020-09-18 19:19:34 ERROR (MainThread) [pyisy.events.websocket] Error during receive Invalid UTF-8 text message
2020-09-18 19:19:34 WARNING (MainThread) [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
2020-09-18 19:19:34 DEBUG (MainThread) [pyisy.events.websocket] Stopping websocket connection.
2020-09-18 19:19:34 INFO (MainThread) [pyisy.events.websocket] PyISY attempting stream reconnect in 0.01s.
2020-09-18 19:19:34 DEBUG (MainThread) [pyisy.events.websocket] Starting websocket connection.
2020-09-18 19:19:35 DEBUG (MainThread) [pyisy.events.websocket] Successfully connected to websocket.
2020-09-18 19:19:35 DEBUG (MainThread) [pyisy.events.websocket] ISY HEARTBEAT: 2020-09-18T19:19:35.106515
```
So it looks like it's at least reconnecting when it sees invalid UTF8, which I guess is good, but individual nodes appear to go offline when that happens?

I'll take a look at the console shortly to see if I can tell what might be the invalid UTF8.

I also plan to report a bug to UDI, but it could take a year for them to fix that!

**Comment by derekatkins** [comment]
Okay, just got to explore a little bit, and ZW003_354 showed up again, with that "micro" character again.  But I don't know why that's stopping ZW010 from reporting.  I don't see any nodes there with an invalid character.  I suppose I could pull down the status again and see; the logs don't say what node is being processed; it would be useful to output the byte-array to the log on decode error.

**Comment by derekatkins** [comment]
Okay, more digging -- during the "outage", I do see ZW003_1 CLITEMP messages of the form:
```txt
2020-09-19 10:43:01 DEBUG (MainThread) [pyisy] ISY Node Control Event: NodeProperty('ZW003_1': control='CLITEMP', value='386', prec='1', uom='17', formatted='38.6°F')
```
However the HA entity isn't being updated from these messages from 10:35 until about 12:00.

On the other hand, I *did* stop receiving CLITEMP messages from ZW010_1 into HA as of 19:19:28 last night (but as I've stated, I do see them in the ISY).

So... what else do you want me to look at?

**Comment by derekatkins** [comment]
FYI, reported to UDI: https://forum.universal-devices.com/topic/30798-invalid-utf-8-in-xml-node-data-on-510-from-a-bogus-z-wave-node/

**Comment by shbatm** [comment]
> Invalid UTF-8 text message

This error is thrown from `aiohttp`'s websocket, which looks like it immediately closes the socket on invalid `UTF-8`. That's not good for us. There's no "ignore" option that I can see in the code that would just let it move on or print the byte array it received, at least that I found in a quick check of the source.

It is reacting and reconnecting, but I'm not sure why you wouldn't be receiving data for that individual device any further. The websocket should be receiving everything from the ISY and as long as the device was loaded into HA at startup, it should know about it and receive the messages.

Depending on how comfortable you are with Python, you can run the beta module in a console and direct the logs somewhere convenient.

```shell
pip3 install pyisy-beta==3.0.0dev14  # install the module
python3 -m pyisy http://url-to-isy:port username password -v >> isylog.txt  # connect websocket and log
```

Then you should be able to grep/filter the log file from another terminal and see if it's getting messages for those devices after they "disconnect". It will print the raw XML before the "ISY Updated Node" or "ISY Node Control Event" lines.

I'll keep looking for anything about why it would stop updating HA, unless the ISY just stops reporting that device to the websocket for some reason.

**Comment by derekatkins** [comment]
Yeah, there is a longstanding (locked) issue about it:  https://github.com/aio-libs/aiohttp/issues/1802 which seems to imply we should use .read() instead of .text().

Question: how does the ISY know what data to send to you?  Do you have to register for the specific node data, or does it send all updates over the websocket?  If you have to register for the nodes that you want, could part of the issue be that it is dying on parsing the list of nodes before it gets to ZW010?  Specifically, we know the bad did is in ZW003, and I would expect that to arrive before ZW010, so maybe HA never "learns" about ZW010 and therefore cannot subscribe to WS notifications to ZW010?
(I'm just grasping here, without knowing the underlying code).

I must admit that I'm a python on-looker.  While I can understand and edit python code, I would definitely not call myself a python programmer (and I *still* can't get beyond the syntactic whitespace).  I've never used pip.  The disconnects seem to happen every 10 second or so!!
```txt
config $ tail -f home-assistant.log | grep -i disconnect
2020-09-19 13:57:10 WARNING (MainThread) [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
2020-09-19 13:57:19 WARNING (MainThread) [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
2020-09-19 13:57:31 WARNING (MainThread) [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
2020-09-19 13:57:41 WARNING (MainThread) [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
```
Let me see if I can pip3 install and play with this with/for you.  Too bad I can't just run it out of the PyISY git clone?

**Comment by derekatkins** [comment]
So I see this:
```txt
2020-09-19 14:04:11 ERROR [pyisy] ISY Could not parse response, poorly formatted XML.: Nodes
```
and then it prints out my ISY device tree.  I see ZW010 in there.  It reconnects, and then later I see similar data to what I saw before:
```txt
2020-09-19 14:06:58 Level 5 [pyisy.events.websocket] ISY Update Received:
<?xml version="1.0"?><Event seqnum="483" sid="uuid:7304"><control>ST</control><action uom="100" prec="0">0</action><node>50 BE 48 1</node><eventInfo></eventInfo><fmtAct>Off</fmtAct></Event>
2020-09-19 14:06:58 DEBUG [pyisy] ISY Updated Node: 50 BE 48 1
2020-09-19 14:06:58 ERROR [pyisy.events.websocket] Error during receive Invalid UTF-8 text message
2020-09-19 14:06:58 WARNING [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
```
I don't see the actual data that was sent that caused the invalid UTF8 message?

**Comment by shbatm** [comment]
> Let me see if I can pip3 install and play with this with/for you. Too bad I can't just run it out of the PyISY git clone?

`pip` is just a python package manager; if you're running this on the same instance/environment as your home assistant instance, you'll have pip already, since that's what HA uses; if not, it should be available just about wherever python is.

> Too bad I can't just run it out of the PyISY git clone?

You may be able to with `python3 __main__.py` run inside of that folder instead of `python3 -m pyisy`. You can also "install" the local clone with `pip install -e /path/to/pyisy`. 

> 2020-09-19 14:04:11 ERROR [pyisy] ISY Could not parse response, poorly formatted XML.: Nodes

This means we have another issue with reading outside of the websocket, since that would come from the initial query of /rest/nodes. I'll look into that, we should be able to fix/print exactly where it's failing.

> I don't see the actual data that was sent that caused the invalid UTF8 message?

The whole `<?xml version="1.0"?>...` line is the "chunk" received by the websocket, so the one event that is crashing it is not even being fed back to `pyisy` from `aiohttp` where we can see the bad character, I need to see how to get `aiohttp` to provide the full error message. That aiohttp issue is similar, but is for regular requests, not for websockets--we use `.content` instead of `.text` for requests. Websockets return a type and a message (https://docs.aiohttp.org/en/stable/websocket_utilities.html#aiohttp.WSMessage). 

The subscription to the ISY is an all or none--there's no picking what it sends you. PyISY takes everything in and only discards things for nodes it knows nothing about--but you should still see the raw socket message printed.

**Comment by shbatm** [comment]
The issue seems to be [here](https://github.com/aio-libs/aiohttp/blob/6a5ab96bd9cb404b4abfd5160fe8f34a29d941e5/aiohttp/http_websocket.py#L384) since the websocket forces the decode of UTF-8 with the default "strict" method, it doesn't allow us to ignore errors, nor does it seem to pass the full stack back to see what the error was.

**Comment by derekatkins** [comment]
Well, that is unfortunate.  :(

**Comment by derekatkins** [comment]
Can we override that method in the class?  ;)

**Comment by bdraco** [comment]
> The issue seems to be [here](https://github.com/aio-libs/aiohttp/blob/6a5ab96bd9cb404b4abfd5160fe8f34a29d941e5/aiohttp/http_websocket.py#L384) since the websocket forces the decode of UTF-8 with the default "strict" method, it doesn't allow us to ignore errors, nor does it seem to pass the full stack back to see what the error was.
> 
> @bdraco See any workarounds I'm missing with `aiohttp`?


Possible options:

 Trap WebSocketError and warn if its WSCloseCode.INVALID_TEXT otherwise raise

 Change out the parser see `aiohttp/client.py` line `proto.set_parser(WebSocketReader(reader, max_msg_size), reader)`

**Comment by derekatkins** [comment]
> Trap WebSocketError and warn if its WSCloseCode.INVALID_TEXT otherwise raise

I don't see how this necessarily helps.  It will keep it from crashing, but might not keep the session open, and it most likely wont let us see what the errant message is/was.  Maybe that's sufficient?

> Change out the parser see `aiohttp/client.py` line `proto.set_parser(WebSocketReader(reader, max_msg_size), reader)`

This might be useful?

I am certainly happy to test, since it appears to be easily reproducible at the moment.  So long as ZW003 continues to report, I am fine ignoring ZW010 for now.

**Comment by shbatm** [comment]
@derekatkins can you please try changing [this line](https://github.com/automicus/PyISY/blob/9965b9280dfd8cc2c8b58357a90df89aa6df1c75/pyisy/events/websocket.py#L238):

```python
_LOGGER.error("Error during receive %s", ws.exception())
```

to this:

```python
_LOGGER.error("Error during receive %s", msg.data, exc_info=True)
```

and see if you get some more information?

If you installed with pip, you can see where it put the module with `pip show pyisy-beta`. If you used the local clone with `pip install -e /path/to/repo` you should be able to just change that copy.

**Comment by derekatkins** [comment]
I made that change and now it says:
```txt
2020-09-19 21:50:50 ERROR [pyisy.events.websocket] Error during receive Invalid UTF-8 text message
NoneType: None
2020-09-19 21:50:50 WARNING [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
```

**Comment by derekatkins** [comment]
So, I upgraded my ISY to 5.2.0 as per the discussion there, but now I get an error from PYISY:
```txt
2020-09-20 13:10:50 DEBUG [pyisy.events.websocket] Starting websocket connection.
2020-09-20 13:10:50 ERROR [pyisy.events.websocket] Unexpected websocket error 500, message='Invalid response status', url=URL('ws://ISY:80/rest/subscribe')
Traceback (most recent call last):
  File "/home/warlord/.local/lib/python3.7/site-packages/pyisy/events/websocket.py", line 226, in websocket
    ssl=self.sslcontext,
  File "/home/warlord/.local/lib/python3.7/site-packages/aiohttp/client.py", line 1012, in __aenter__
    self._resp = await self._coro
  File "/home/warlord/.local/lib/python3.7/site-packages/aiohttp/client.py", line 738, in _ws_connect
    headers=resp.headers)
aiohttp.client_exceptions.WSServerHandshakeError: 500, message='Invalid response status', url=URL('ws://ISY:80/rest/subscribe')
```
For kicks, let me try https instead of http.  Okay, that seems to be working; so now I need to wait for that bogus node to obtain some data.  Who knows when that will happen.  But I also appear to need to change my integration to use SSL, since WS on port 80 doesn't appear to work anymore.

**Comment by derekatkins** [comment]
quick update -- restarted HA and everything is connected and working.  Need to wait for that bogus node to show again..

**Comment by derekatkins** [comment]
... and the bogus node reappeared, and 5.2.0 is still sending ISO and not UTF8.  HA is still flapping. Running PYISY directly I see:
```txt
2020-09-24 08:31:37 ERROR [pyisy] ISY Could not parse response, poorly formatted XML.: Nodes
[ Node Tree Elided here ]
2020-09-24 08:31:37 INFO [__main__] Total Loading time: 30.02s
2020-09-24 08:31:37 DEBUG [pyisy.events.websocket] Starting websocket connection.
2020-09-24 08:32:37 DEBUG [pyisy.events.websocket] Websocket Server Not Ready.
2020-09-24 08:32:37 DEBUG [pyisy.events.websocket] Stopping websocket connection.
2020-09-24 08:32:37 INFO [pyisy.events.websocket] PyISY attempting stream reconnect in 0.01s.
2020-09-24 08:32:37 DEBUG [pyisy.events.websocket] Starting websocket connection.
2020-09-24 08:32:42 DEBUG [pyisy.events.websocket] Successfully connected to websocket.
```
Notice the 60-second delay in there?  Then I see, later:
```txt
2020-09-24 08:32:49 DEBUG [pyisy] ISY Node Control Event: NodeProperty('ZW003_308': control='ALARM', value='0', prec='0', uom='25', formatted='0')
2020-09-24 08:32:49 ERROR [pyisy.events.websocket] Error during receive Invalid UTF-8 text message
NoneType: None
2020-09-24 08:32:49 WARNING [pyisy.events.websocket] Websocket disconnected unexpectedly with code: 1006
2020-09-24 08:32:49 DEBUG [pyisy.events.websocket] Stopping websocket connection.
2020-09-24 08:32:49 INFO [pyisy.events.websocket] PyISY attempting stream reconnect in 0.01s.
2020-09-24 08:32:49 DEBUG [pyisy.events.websocket] Starting websocket connection.
2020-09-24 08:32:59 DEBUG [pyisy.events.websocket] Successfully connected to websocket.
```
So it is still failing on that ISO character in the XML response.

**Comment by derekatkins** [comment]
> You may be able to with `python3 __main__.py` run inside of that folder instead of `python3 -m pyisy`. You can also "install" the local clone with `pip install -e /path/to/pyisy`.

I tried this, but I get:
```txt
Traceback (most recent call last):
  File "__main__.py", line 17, in <module>
    from . import ISY
ImportError: attempted relative import with no known parent package
```
Doesn't matter where I run it from, I still get this error.  I can try to install first, but there's got to be a way to tell it to run-from-source?

**Comment by derekatkins** [comment]
> > 2020-09-19 14:04:11 ERROR [pyisy] ISY Could not parse response, poorly formatted XML.: Nodes
> 
> This means we have another issue with reading outside of the websocket, since that would come from the initial query of /rest/nodes. I'll look into that, we should be able to fix/print exactly where it's failing.

I tracked this down to this code in `nodes/__init__.py` in `async def update(self, wait_time=0, xml=None):`
```txt
        try:
            xmldoc = minidom.parseString(xml)
        except XML_ERRORS:
            _LOGGER.error("%s: Nodes <414>", XML_PARSE_ERROR)
            return False
```
(I added the 414 in there to track down where it was occurring).  The fact that it's on `/status` makes sense, as we know that has the ISO character in there.  Is there a way to tell minidom.parseString to ignore UTF8 errors?

**Comment by shbatm** [comment]
> The fact that it's on `/status` makes sense, as we know that has the ISO character in there. Is there a way to tell minidom.parseString to ignore UTF8 errors?

This part has been fixed in #125 and #126 already. I haven't had a chance to fix the v3 websocket issue, but everything for v2 (which is what HA uses) will be fixed in the next release.

**Comment by derekatkins** [comment]
It's not been fixed in #126 -- I am running that branch in testing and still seeing the issue.

**Comment by derekatkins** [comment]
@shbatm Has the send_node_command issue been fixed in current HA release?  If so, I can downgrade back to version 2 once it's "released".  Although right now I am testing v3.

**Comment by derekatkins** [comment]
Answering my own question (after downgrading back to v3) -- yes, it has.  So I'll run with v2 for now until you tell me you think you're ready to test v3.  Of course, I removed the bogus node from my ISY earlier today so of course everything is working for now.  I'll keep an eye out for if (when?) the bogus node comes back.

Thanks!

**Comment by derekatkins** [comment]
FYI, as I just mentioned in https://github.com/automicus/PyISY/pull/125, the fix didn't fix the problem.  I got another (different) error code in the logs when the bogus node reappeared and then the ISY stopped sending events to HA:
```txt
Control Event: ZW003_1 NodeProperty('CLIHUM': value='66' prec='0' uom='22' formatted='66%')
2020-10-03 09:25:13 DEBUG (Thread-7) [homeassistant.components.isy994] ISY Updated Program: 003D
2020-10-03 09:25:16 ERROR (Thread-7) [root] Uncaught thread exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.8/site-packages/pyisy/events.py", line 220, in watch
    events = event_reader.read_events(POLL_TIME)
  File "/usr/local/lib/python3.8/site-packages/pyisy/eventreader.py", line 53, in read_events
    events.append(body.decode("utf-8"))
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb5 in position 177: invalid start byte
2020-10-03 09:38:23 INFO (SyncWorker_7) [homeassistant.components.isy994] ISY Request: https://smarthome.ihtfp.org:443/rest/vars/set/1/3/600
2020-10-03 09:38:24 DEBUG (SyncWorker_7) [homeassistant.components.isy994] ISY Response Received
2020-10-03 09:38:24 DEBUG (SyncWorker_7) [homeassistant.components.isy994] ISY set variable: 1.3
2020-10-03 09:38:24 INFO (SyncWorker_7) [homeassistant.components.isy994] ISY Request: https://smarthome.ihtfp.org:443/rest/vars/get/1
2020-10-03 09:38:24 DEBUG (SyncWorker_7) [homeassistant.components.isy994] ISY Response Received
2020-10-03 09:38:24 INFO (SyncWorker_7) [homeassistant.components.isy994] ISY Request: https://smarthome.ihtfp.org:443/rest/vars/get/2
2020-10-03 09:38:24 DEBUG (SyncWorker_7) [homeassistant.components.isy994] ISY Response Received
2020-10-03 09:38:24 INFO (SyncWorker_7) [homeassistant.components.isy994] ISY Loaded Variables
```

**Comment by derekatkins** [comment]
I was going to try to test the 2.0 branch in the way you suggest testing the 3.0 branch, but there is no __main__.py in the 2.0 branch.  :(

**Comment by shbatm** [comment]
Update on v3/Websockets -- After discussion on `aiohttp`'s repo and digging further into the Websocket API and other implementations, everything is pretty unambiguous about breaking the socket connection on invalid UTF-8 messages and sending a 1007 close message. The issue with this, as you've seen, is a reconnect is not instantaneous and upon reconnection to the ISY, it will send a full system status dump again, which takes time to process.

Unfortunately, I'm at the point where I will make sure we handle any invalid UTF-8 characters in the initial query of the ISY (calls to `/rest/nodes` and `/rest/status`), but I'm not in support of trying to overload the `aiohttp` functions to handle bad characters from the ISY. The point of moving to v3 is to get away from the custom socket implementations and use standard modules for the communications.

tl;dr: I don't support adding workarounds for the v3 Websocket, ISY needs to fix the invalid character being sent because it does not conform to the Websocket standard. 



> I was going to try to test the 2.0 branch in the way you suggest testing the 3.0 branch, but there is no `__main__.py` in the 2.0 branch. :(

v2 has an `examples` folder which has a `test_connection.py` script in it you can use in a similar fashion to v3's module code. You'll need to edit the file to add your connection details.

**Comment by derekatkins** [comment]
> tl;dr: I don't support adding workarounds for the v3 Websocket, ISY needs to fix the invalid character being sent because it does not conform to the Websocket standard.

Thanks for the update.  I don't blame you, but I do hope you wait to push v3 until UDI has fixed this bug (which they acknowledged in the 5.3.0 release thread and said would be in the "next build").
> > I was going to try to test the 2.0 branch in the way you suggest testing the 3.0 branch, but there is no `__main__.py` in the 2.0 branch. :(
> 
> v2 has an `examples` folder which has a `test_connection.py` script in it you can use in a similar fashion to v3's module code. You'll need to edit the file to add your connection details.

I am actually running an updated v2 in HA, manually applying your v2 changes, and even with the ISY returning non-UTF8, it's working perfectly.  So I can confirm that your v2 changes work.  I saw you merged and I am waiting for this change to get into the HA release (I suspect 0.117?)

**Comment by shbatm** [comment]
> I saw you merged and I am waiting for this change to get into the HA release (I suspect 0.117?)

Yes, should be in the next major version release.


> I don't blame you, but I do hope you wait to push v3 until UDI has fixed this bug (which they acknowledged in the 5.3.0 release thread and said would be in the "next build").

I saw Chris' response I will keep an eye on the build posts--the plan was to test V3 until the end of the year using the [hacs-isy994](https://github.com/shbatm/hacs-isy994) `custom_component` and then if no major issues, migrate to Home Assistant Core. If UDI still hasn't resolved the bug by the end of the year, I may just do a flip-flop and push a v2 version to HACS and promote the V3 version in Core; that will give you an avenue to wait until they fix the problem before upgrading.

**Comment by derekatkins** [comment]
> I saw Chris' response I will keep an eye on the build posts--the plan was to test V3 until the end of the year using the [hacs-isy994](https://github.com/shbatm/hacs-isy994) `custom_component` and then if no major issues, migrate to Home Assistant Core. If UDI still hasn't resolved the bug by the end of the year, I may just do a flip-flop and push a v2 version to HACS and promote the V3 version in Core; that will give you an avenue to wait until they fix the problem before upgrading.

This is a fair approach (although I will probably just install the custom_component by hand instead of using HACS).  I'm just asking for v2 to remain supported until this bug is fixed by UDI.  Thanks!

**Comment by shbatm** [comment]
@derekatkins Did you ever see if this was updated on UDI's side?

**Comment by derekatkins** [comment]
@shbatm I believe it was updated in 5.3.1 (although I am still running 5.2.0)

**Comment by shbatm** [comment]
Fixed upstream, not able to fix websockets here due to standards. Closing for now.
